### PR TITLE
Fix/498 buffer protocol

### LINKED ISSUE #498: Potential memory bug for memory.buffer
Not sure this really is an issue in practice, but I still wanted to raise it. At the moment wasmer does not set the owner pointer of the buffer view (`obj`, [relevant location](https://github.com/wasmerio/wasmer-python/blob/953a17ad7666ea623a7b08419e347a76ec9107e7/packages/api/src/memory/buffer.rs#L100)). 

My understanding is that in principle this could lead to invalid memory access, if the buffer outlives the reference to the wasmer objects it was created from. In practice this should be a non issue as numpy, and I belief also memoryview, explictly retrain a reference to the object used to get the buffer view. However, this additional saftey is not required by Python and therefore different consumers of the buffer view may encounter memory bugs.

To correctly track the owner of the memory, it should be enough to change the [linked line](https://github.com/wasmerio/wasmer-python/blob/953a17ad7666ea623a7b08419e347a76ec9107e7/packages/api/src/memory/buffer.rs#L100) to

```rust
use pyo3::ffi::Py_IncRef;

(*view).obj = slf.as_ptr();
Py_IncRef((*view).obj);
```

This change would follow the [documented buffer protocol](https://docs.python.org/3.9/c-api/typeobj.html#c.PyBufferProcs.bf_getbuffer): "Set view->obj to exporter and increment view->obj." No change of the release method is required, as Python will decrement the ref count according to the [docs](https://docs.python.org/3.9/c-api/typeobj.html#c.PyBufferProcs.bf_releasebuffer).

A small caveat: I only started to look into the buffer protocol recently, so there is a good chance I misunderstood some of the contracts involved :).

### PR DESCRIPTION
**Pull request by chmp**
See #498 for details and discussion of the changes.

Close #498.

### CODE REVIEW

**Comment by Hywan** [comment]
I've fixed the CI, and tests are running again.

File: tests/test_memory.py
Comment (Hywan) [inline tests/test_memory.py:302]: To be sure the memory instance is deleted, you have to either you `del` (but it's not sure the GC will collect it) or (better) force the GC to collect it.

File: tests/test_memory.py
Comment (Hywan) [inline tests/test_memory.py:307]: What does it check?

File: tests/test_memory.py
Comment (chmp) [reply tests/test_memory.py:302]: You're right. I'll add the `gc.collect` part. I think CPython will always collect eagerly, if there are no cycles (as is the case here), but probably this does not apply to other interpreters.

The `del` call is not required, as the `memory.buffer` is only used as a temporary inside the line. But I guess making it two lines and explicitly deteting the ref to buffer is more self-documenting.

File: tests/test_memory.py
Comment (chmp) [reply tests/test_memory.py:307]: The idea was to write to memory, that is only valid to write to, if the ownership semantics are correctly expressed. Probably, not a very good test, as it is intederministic in nature outside of a memory debugger such as valgrind. I.e., depending on your system, it could pass even, though you completely destroy non-owned memory

At least on my system, it immediately segfeaulted with the previous ownership setup (`(*view).obj = ptr::null_mut();`). 

TBH. I'm not sure it makes sense to keep it. I'll add a docstring explaining the motivation behind the test, but I'm happy to remove it. In any case, it was helpful for me to verify that there was indeed an issue.

**Comment by chmp** [comment]
Re tests: they now fail on aarch64 with 

```
just build-wheel compiler-singlepass python3.9 aarch64-unknown-linux-gnu
sh: 1: just: not found
```

Since I did not modify anything related, I have not even the remotest idea how I would, or even could, fix it.

File: tests/test_memory.py
Comment (Hywan) [reply tests/test_memory.py]: > The del call is not required, as the memory.buffer is only used as a temporary inside the line. But I guess making it two lines and explicitly deteting the ref to buffer is more self-documenting.

Exactly, that was my idea. Thanks for reading in my minds :-p.

File: tests/test_memory.py
Comment (Hywan) [reply tests/test_memory.py:325]: I'm happy to keep the tests to avoid regression. Please add a documentation explaining what it showcases even if it doesn't test exactly what we want. That's a valid regression test after all :-)!

**Comment by Hywan** [comment]
bors try

**Comment by chmp** [comment]
Perfect. Tests and build passsed :). I think your comments should be addressed int the current state or should I change I anything else?

**Comment by Hywan** [comment]
It looks to me :-). Nothing is necessary. I'm waiting on the CI to give another verdict before merging.

**Review by Hywan — APPROVED** [review]
bors r+

**Comment by chmp** [comment]
> It looks to me :-). Nothing is necessary. I'm waiting on the CI to give another verdict before merging.

Cool. I have to say I did enjoy this puzzle box.

**Comment by Hywan** [comment]
Tests are OK but bors is in a bad mood. Let's by-pass him :-).

**Pull request merged**
